lib/approval_request_listener.rb

```diff
@@ -0,0 +1,64 @@
+require "manageiq-messaging"
+
+class ApprovalRequestListener
+  SERVICE_NAME = "platform.approval".freeze
+  CLIENT_AND_GROUP_REF = "catalog-api-worker".freeze
+
+  class ApprovalRequestNotFound < StandardError; end
+
+  attr_accessor :messaging_client_options, :client
+
+  def initialize(messaging_client_options = {})
+    self.messaging_client_options = default_messaging_options.merge(messaging_client_options)
+  end
+
+  def run
+    Thread.new { subscribe_to_approval_updates }
+  end
+
+  def subscribe_to_approval_updates
+    self.client = ManageIQ::Messaging::Client.open(messaging_client_options)
+
+    client.subscribe_messages(
```

Comment:
You should subscribe to events rather than messages, where you also get the event name.
The approval-api send events with the following types:
```
  EVENT_REQUEST_STARTED  = 'request_started'.freeze
  EVENT_REQUEST_FINISHED = 'request_finished'.freeze
  EVENT_APPROVER_GROUP_NOTIFIED = 'approver_group_notified'.freeze
  EVENT_APPROVER_GROUP_FINISHED = 'approver_group_finished'.freeze
```

Reply:
@bzwei This is outside the scope/intent of this PR, but should we be doing this for the `ServiceOrderListener` as well? Not the same events, obviously, but is the intent that the publisher of messages send events? My understanding was that the `.subscribe_messages(:service => SERVICE_NAME)` was the preferred implementation. Since we control the messages that are being sent to the `ServiceOrderListener` on the topology side, though, I could change them to send events instead.

Reply:
You will have to use `publish_event` / `subscribe_event`, 'publish_message` / `subscribe_message` in pairs. Kafka essentially is an event type broker. We provide a flavor of message in order to fully implement the `manage-messaging` interface. Event is the recommendation unless you have a specific reason to use message.

Reply:
Topic is another name for event.